### ISSUE TITLE
OnDeserialized hook is not called when a class have no @deserialize field

### ISSUE THREAD

**Issue by 1ambda** [weichx/cerialize#16]
Assume that we have models deserializing Github API responses like

```typescript
export class GithubEvent {
  @deserializeAs("id") public event_id: string;
  @deserializeAs("type") public event_type: string;
  @deserializeAs("created_at") public created_at: string;
  public actor: string; /** actor.login */
  public repo: string; /** owner/repo_name */

  public static OnDeserialized(instance : GithubEvent, json : any) : void {
    if (!_.isEmpty(json) && !_.isEmpty(json.actor))
      instance.actor = json.actor.login;

    if (!_.isEmpty(json) && !_.isEmpty(json.repo))
      instance.repo = json.repo.name;
  }
}

@inheritSerialization(GithubEvent)
export class GithubRepositoryEvent extends GithubEvent {
  public static EVENT_TYPE = "RepositoryEvent";
}

export class GithubForkEventPayload {
  /** even if ForkEventPayload doesn't need to have action field,
   *  we need it otherwise `OnDeserialized` is not called
   */
  @deserializeAs("action") public action: string;
  public forkee_id: string;
  public forkee_name: string;
  public forkee_full_name: string;
  public forkee_url: string;
  public forkee_language: string;

  public static OnDeserialized(instance: GithubForkEventPayload, payload: any): void {
    if (_.isEmpty(payload)) return;

    if (!_.isEmpty(payload.forkee)) {
      let forkee = payload.forkee;
      instance.forkee_id = forkee.id;
      instance.forkee_name = forkee.name;
      instance.forkee_full_name = forkee.full_name;
      instance.forkee_url = forkee.html_url;
      instance.forkee_language = forkee.language;
    }
  }
}
```

```typescript
// spec
let forkEvent1 = {
  "id": "3475819128",
  "type": "ForkEvent",
  "actor": {
    "login": "1ambda",
  },
  "repo": {
    "id": 41002005,
    "name": "weichx/cerialize",
    "url": "https://api.github.com/repos/weichx/cerialize"
  },
  "payload": {
    "forkee": {
      "id": 48599725,
      "name": "cerialize",
      "full_name": "1ambda/cerialize",
      "private": false,
      "html_url": "https://github.com/1ambda/cerialize",
      "language": "TypeScript",
    }
  },
  "created_at": "2015-12-26T04:47:20Z"
};

describe("ForkEvent", () => {
  it("should be deserialized", () => {
    let raw = forkEvent1;
    let forkEvent = GithubForkEvent.deserialize(GithubForkEvent, raw);

    expect(forkEvent.event_id).toEqual(raw.id);
    expect(forkEvent.event_type).toEqual(raw.type);
    expect(forkEvent.event_type).toEqual(GithubForkEvent.EVENT_TYPE);
    expect(forkEvent.actor).toEqual(raw.actor.login);
    expect(forkEvent.repo).toEqual(raw.repo.name);
    expect(forkEvent.created_at).toEqual(raw.created_at);

    let payload = forkEvent.payload;

   // the below tests fail if we remove `@deserialize` field in `GithubForkEvenyPayload`
    expect(payload.forkee_id).toEqual(raw.payload.forkee.id);
    expect(payload.forkee_name).toEqual(raw.payload.forkee.name);
    expect(payload.forkee_full_name).toEqual(raw.payload.forkee.full_name);
    expect(payload.forkee_url).toEqual(raw.payload.forkee.html_url);
    expect(payload.forkee_language).toEqual(raw.payload.forkee.language);
  });
});
```

In this case, if we remove `@deserializeAs("action") public action: string;` field, `OnDeserialized` hook in `GithubForkEventPayload` class will be not called because [metadataArray](https://github.com/weichx/cerialize/blob/a45c93190ca488848474734e8242a9c37bfbecfd/src/serialize.ts#L408) is empty. 

So I think `OnDeserialized` hook logic should be called before checking  [metadataArray](https://github.com/weichx/cerialize/blob/a45c93190ca488848474734e8242a9c37bfbecfd/src/serialize.ts#L408) in [deserializeObject](https://github.com/weichx/cerialize/blob/a45c93190ca488848474734e8242a9c37bfbecfd/src/serialize.ts#458) and [desrializedObjectInto](https://github.com/weichx/cerialize/blob/a45c93190ca488848474734e8242a9c37bfbecfd/src/serialize.ts#L458) functions. 

I already created PR for `OnDeserialized` in https://github.com/weichx/cerialize/pull/15. If you don't mind, I would like to fix this bug on another PR after #15 is merged as this is the different issue.

**Comment by weichx** [comment]
Interesting use case, I think you are right that the hook should be called but I would like to find a nice way to do so at parse-time instead of run time. Since we use `@inheritSerialization` anyway, perhaps the right time for this check is inside the `@inheritSerialization` attribute function body. I'll play with this and see what I can come up with.

**Comment by weichx** [comment]
https://github.com/weichx/cerialize/pull/17 should solve this issue for the non inherited case

**Comment by weichx** [comment]
I'm closing this and addressing inherited serialization separately since I think that is an ongoing discussion and the root issue here has been fixed.

**Comment by 1ambda** [comment]
Thanks :)
